### PR TITLE
Fix compile errors for configuration gShield

### PR DESCRIPTION
**Pull request by lrepasi**
I used the othermill configuration as an example of how to fix the gShield configuration.

### CODE REVIEW

**Comment by justinclift** [comment]
This is for getting the `edge-preview` branch working on the Due again, yeah? :smile:

If so, have you taken a look at @crispy1989's branch here?

&nbsp; &nbsp; https://github.com/crispy1989/g2/tree/edge-preview-arduino-due-v1

That was an initial attempt a few weeks ago and works to the level of "seems ok" but hasn't undergone any in-depth testing at all.  Might be useful to compare that and yours to see if there are bits which need merging across. :smile:

**Comment by lrepasi** [comment]
Hi Justin,

A couple of days ago I made the code changes but did not commit them. At that time I found your changes. The only difference between your code and my code was that you kept the code for esc_spindle. After seeing your code I decided to put esc_spindle back into my code. I’m still not sure if it needs to be there. If it doesn’t it’s easily removed in the future. My main goal is to have the Arduino Due board compile cleanly and it does. At this point your code and my code are basically the same.

In your changes I think all you needed to do is update the settings file and it would have compiled.

File: g2core/board/ArduinoDue/board_gpio.cpp [hunk 1]
```diff
@@ -56,39 +56,39 @@
...
+gpioDigitalInputPin<IRQPin<Motate::kInput5_PinNumber>>  din5  {DI5_ENABLED,  DI5_POLARITY,  5, DI5_EXTERNAL_NUMBER, Motate::kPinInterruptOnChange|Motate::kPinInterruptPriorityHigh};
+gpioDigitalInputPin<IRQPin<Motate::kInput6_PinNumber>>  din6  {DI6_ENABLED,  DI6_POLARITY,  6, DI6_EXTERNAL_NUMBER, Motate::kPinInterruptOnChange|Motate::kPinInterruptPriorityHigh};
+gpioDigitalInputPin<IRQPin<Motate::kInput7_PinNumber>>  din7  {DI7_ENABLED,  DI7_POLARITY,  7, DI7_EXTERNAL_NUMBER, Motate::kPinInterruptOnChange|Motate::kPinInterruptPriorityHigh};
+gpioDigitalInputPin<IRQPin<Motate::kInput8_PinNumber>>  din8  {DI8_ENABLED,  DI8_POLARITY,  8, DI8_EXTERNAL_NUMBER, Motate::kPinInterruptOnChange|Motate::kPinInterruptPriorityHigh};
+gpioDigitalInputPin<IRQPin<Motate::kInput9_PinNumber>>  din9  {DI9_ENABLED,  DI9_POLARITY,  9, DI9_EXTERNAL_NUMBER, Motate::kPinInterruptOnChange|Motate::kPinInterruptPriorityHigh};
+gpioDigitalInputPin<IRQPin<Motate::kSocket2_SPISlaveSelectPinNumber>> din10 {DI10_ENABLED, DI10_POLARITY, 10, DI10_EXTERNAL_NUMBER, Motate::kPinInterruptOnChange|Motate::kPinInterruptPriorityHigh};
```
Comment (giseburt) [inline g2core/board/ArduinoDue/board_gpio.cpp]: This was a copy-paste of a hack on the G2v9k board to steal an unused slave select for an extra input. Please use `kInput10_PinNumber` instead:
```suggestion
gpioDigitalInputPin<IRQPin<Motate::kInput10_PinNumber >> din10 {DI10_ENABLED, DI10_POLARITY, 10, DI10_EXTERNAL_NUMBER, Motate::kPinInterruptOnChange|Motate::kPinInterruptPriorityHigh};
```

File: g2core/board/ArduinoDue/board_gpio.h [hunk 2]
```diff
@@ -68,7 +78,7 @@ extern gpioDigitalInputPin<IRQPin<Motate::kInput6_PinNumber>>  din6;
 extern gpioDigitalInputPin<IRQPin<Motate::kInput7_PinNumber>>  din7;
 extern gpioDigitalInputPin<IRQPin<Motate::kInput8_PinNumber>>  din8;
 extern gpioDigitalInputPin<IRQPin<Motate::kInput9_PinNumber>>  din9;
-// extern gpioDigitalInputPin<IRQPin<Motate::kInput10_PinNumber>> din10;
+extern gpioDigitalInputPin<IRQPin<Motate::kSocket2_SPISlaveSelectPinNumber>> din10;
```
Comment (giseburt) [inline g2core/board/ArduinoDue/board_gpio.h]: Same here:
```suggestion
extern gpioDigitalInputPin<IRQPin<Motate::kInput10_PinNumber>> din10;
```

File: g2core/settings/settings_shapeoko2.h [hunk 3]
```diff
@@ -156,70 +157,91 @@
...
+    IO_UNAVAILABLE   // input/output is missing/used/unavailable
+    IO_DISABLED      // input/output is disabled
+    IO_ENABLED       // input/output enabled
+
+  DOn_POLARITY:
+    IO_ACTIVE_LOW    aka NORMALLY_OPEN
+    IO_ACTIVE_HIGH   aka NORMALLY_CLOSED
+
 */
 // Xmin on v9 board
```
Comment (giseburt) [inline g2core/settings/settings_shapeoko2.h]:
```suggestion
```

File: g2core/settings/settings_shapeoko2.h [hunk 4]
```diff
@@ -156,70 +157,91 @@
...
-#define DI1_MODE                    NORMALLY_CLOSED
+#define DI1_ENABLED                 IO_ENABLED
+#define DI1_POLARITY                NORMALLY_CLOSED
 //#define DI1_ACTION                  INPUT_ACTION_STOP
-#define DI1_ACTION                  INPUT_ACTION_NONE
-#define DI1_FUNCTION                INPUT_FUNCTION_LIMIT
+#define DI1_ACTION                  INPUT_ACTION_LIMIT
 
 // Xmax
```
Comment (giseburt) [inline g2core/settings/settings_shapeoko2.h]: These comments aren’t accurate and can be snipped.
```suggestion
// Xmax
```

File: g2core/settings/settings_shapeoko2.h [hunk 5]
```diff
@@ -156,70 +157,91 @@
...
+#define DI8_ENABLED                 IO_DISABLED
+#define DI8_POLARITY                IO_ACTIVE_LOW
 #define DI8_ACTION                  INPUT_ACTION_NONE
-#define DI8_FUNCTION                INPUT_FUNCTION_NONE
 
 // Hardware interlock input
-#define DI9_MODE                    IO_MODE_DISABLED
+#define DI9_ENABLED                 IO_DISABLED
+#define DI9_POLARITY                IO_ACTIVE_LOW
 #define DI9_ACTION                  INPUT_ACTION_NONE
```
Comment (giseburt) [inline g2core/settings/settings_shapeoko2.h:238]: And we should add input 10 back in:
```suggestion
#define DI9_ENABLED                 IO_DISABLED
#define DI9_POLARITY                IO_ACTIVE_LOW
#define DI9_ACTION                  INPUT_ACTION_NONE

#define DI10_ENABLED                 IO_DISABLED
#define DI10_POLARITY                IO_ACTIVE_LOW
#define DI10_ACTION                  INPUT_ACTION_NONE

```

File: g2core/settings/settings_shapeoko2.h [hunk 6]
```diff
@@ -156,70 +157,91 @@
...
-#define DI9_FUNCTION                INPUT_FUNCTION_NONE
```
Comment (giseburt) [inline g2core/settings/settings_shapeoko2.h:225]: I’m not sure what order you’ll see my comments, so read this part as if it came in first: Thank you!

Your time doing this is appreciated, and my comments are really just picky notes, not flaws.

One other thing to note, for anyone using this, is the `kOutput`N`_PinNumber` are all set to `-1` in `g2core/board/ArduinoDue/gShield-pinout.h`, so they won’t be functional until you assign them numbers of the Due pins.

Also, in the same file, at the top, `OUTPUT1_PWM` etc. are all defined as `0`, but most of them could be set to `1` to enable PWM output functionality. You have to be careful about what pins you choose and what timers they’re on, as you may accidentally entangle to outputs that are on the same timer. That’s a chip-level thing, and not something we can control.

It appears that pins 8-12 are not used by the gShield for other things, and are already PWM pins, so we could possibly use those.

**Review by giseburt — CHANGES_REQUESTED** [review]
Thank you for this. I have requested just a few little changes, then I’ll merge this in.

If you can’t do them, that’s fine, but please let me know.

File: g2core/settings/settings_shapeoko2.h (hunk 6 above)
Comment (lrepasi) [reply g2core/settings/settings_shapeoko2.h:225]: I made the changes that you requested and committed them. That's the easy part.

The hard part is I only have 5 days experience with git and github. I don't know how to modify a pull request. Can you point me in the right direction?

Comment (MitchBradley) [reply g2core/settings/settings_shapeoko2.h:225]: If you push your commits to the branch that you used to create the original pull request, github will automatically add that to the pending pull request.

Comment (lrepasi) [reply g2core/settings/settings_shapeoko2.h:225]: I did not know that commits on the same branch subsequent to a pull request contribute to the pull request. I committed a change prior to the commit for your requested changes. You may or may not want that commit.

Is there anything that I should do?

Comment (giseburt) [reply g2core/settings/settings_shapeoko2.h:225]: No worries, that looks great. The only other commit showing is the changes to Atmel Studio files. We’re deprecating support for that, but it doesn’t hurt if someone maintains those files for now. :)

Thank you!

**Review by giseburt — APPROVED** [review]

**Comment by justinclift** [comment]
@breiler The `edge-preview` branch should compile ok for the Due now. :smile:

**Comment by lumascet** [comment]
@lrepasi do you know how to activate the spindle pwm control? as the default settings do not work for the spindle control, I tried  to configure the pins 11 12 and 13 (as the gshield uses these) but i cant get it working.

**Pull request merged**